src/datasets/arrow_dataset.py

```diff
@@ -2654,6 +2654,26 @@ def to_pandas(
                 for offset in range(0, len(self), batch_size)
             )
 
+    def add_column(self, name: str, column: Union[list, np.array]):
+        """Add column to Dataset.
+
+        .. versionadded:: 1.6
+
+        Args:
+            name (str): Column name.
+            column (list or np.array): Column data to be added.
+
+        Returns:
+            :class:`Dataset`
+        """
+        column_table = InMemoryTable.from_pydict({name: column})
+        # Concatenate tables horizontally
+        table = ConcatenationTable.from_tables([self._data, column_table], axis=1)
+        # Update features
+        info = self.info
+        info.features = Features.from_arrow_schema(table.schema)
+        return Dataset(table, info=info)
```

Comment:
@albertvillanova Metadata needs to be updated since this function modifies the features of the dataset.

```suggestion
        table = update_metadata_with_features(table, info.features)
        return Dataset(table, info=info)
```

Reply:
I have finally reused
```python
info = copy.deepcopy(self.info)
```
as suggested by @lhoestq.

Reply:
But this line doesn't actually update the metadata. For instance, this can be verified with:
```python
>>> from datasets import load_dataset
>>> sst = load_dataset("sst", split="train")
>>> sst1 = sst.add_column("my_col", list(range(len(sst))))
>>> print(sst1._data.schema.metadata) # should contain "my_col", but it doesn't
OrderedDict([(b'huggingface', b'{"info": {"features": {"sentence": {"dtype": "string", "id": null, "_type": "Value"}, "label": {"dtype": "float32", "id": null, "_type": "Value"}, "tokens": {"dtype": "string", "id": null, "_type": "Value"}, "tree": {"dtype": "string", "id": null, "_type": "Value"}}}}')])
```

Reply:
Metadata stored in the schema is just a redundant information regarding the feature types.
It is used when calling `Dataset.from_file` to know which feature types to use.
These metadata are stored in the schema of the pyarrow table by using `update_metadata_with_features`.
However this something that's almost never tested properly so we should find a way to always make sure that the metadata are synced with the actual feature types in self.info

In your case you just need to add this line @albertvillanova , thanks @mariosasko for pointing that out
```python
table = update_metadata_with_features(table, info.features)
```

Reply:
I have created issue #2271 in relation with this.